README.md
# Kapow!

An http error library


[![Build Status](https://secure.travis-ci.org/bobholt/kapow.png)](http://travis-ci.org/bobholt/kapow)

Kapow! consolidates boilerplate HTTP information into a single easy-to-use
library. You tell Kapow! that you want an error with a particular httpStatus
code, and it fills in the basics for you. It defines and returns a
`RequestError` object based on JavaScript's native `Error` object.

## Installation

`npm install kapow`

## Usage

### Basics:

To get an JavaScript Error object suitable for throwing or passing along, give
Kapow! a status code, a message, and arbitrary data you want to include:

```js
const Kapow = require('kapow');

var code = 404;
var message = 'Ain\'t nobody here here but us chickens.';
var data = {a: 1};
var err = Kapow(code, message, data);

err;
// Returns:
//  { [RequestError: Ain't nobody here but us chickens.]
//    httpStatus: '404',
//    title: 'Not Found',
//    message: 'Ain\'t nobody here but us chickens.',
//    name: 'RequestError',
//    data: { a: 1 } }
```

All arguments to Kapow! are optional. Kapow! will try to figure out sensible
defaults based on the information you did pass in.

This error can be either thrown immediately or passed on through your
application to eventually send the user an informative HTTP response.

### Wrapping an Existing Error:

If your app hands you an error, and you just want to decorate it, you can
use the static `wrap` method on Kapow!:

```js
const Kapow = require('kapow');

somePromise.then().catch(function(err) {
  Kapow.wrap(err, 404, 'Ain\'t nobody here at all.');
});
```

### Changing Your Error:

Once you get the RequestError object back, you can augment it however you like:

```js
const Kapow = require('kapow');

var err = Kapow(403, 'Don\'t you come back no more');

err.title = 'Hit the Road Jack';

err;

// Returns:
//  { [RequestError: Don't you come back no more]
//    httpStatus: '403',
//    title: 'Hit the Road Jack',
//    message: 'Don\'t you come back no more',
//    name: 'RequestError',
//    data: undefined }
```

### Don't Like Numbers?

You can call Kapow with the camelized version of the IETF's official error.
See [Kapow!'s complete error code list](https://github.com/bobholt/kapow/blob/master/lib/error_codes.js) for all supported codes:

```js
const Kapow = require('kapow');

var err = Kapow('forbidden', 'Fie! Forsooth Begone!');

err;

// Returns:
//  { [RequestError: Fie! Forsooth Begone]
//    httpStatus: '403',
//    title: 'Forbidden',
//    message: 'Fie! Forsooth Begone',
//    name: 'RequestError',
//    data: undefined }
```


CHANGELOG
v1.0.0:
  date: 2015-02-27
  changes:
    - Initial release.


test/index.js
describe('KapowAll', function() {
  require('./unit');
});


test/unit/index.js
const expect = require('chai').expect;
const _ = require('lodash');

const Kapow = require('../../');
const errorCodes = require('../../lib/error_codes');

describe('Kapow', function () {
  describe('lib', function() {
    require('./lib/camelize');
    require('./lib/error_codes');
    require('./lib/partial');
    require('./lib/request_error');
    require('./lib/wrap');
  });

  it('should return a function', function() {
    expect(Kapow).to.be.a('function');
  });

  it('should return an error object when invoked with the new keyword', function() {
    expect(new Kapow()).to.be.instanceof(Error);
  });

  it('should not require the `new` keyword', function() {
    expect(Kapow()).to.be.instanceof(Error);
  });

  it('should have a static method `wrap`', function() {
    expect(Kapow.wrap).to.be.a('function');
  });

  it('should have all error codes as prototype methods', function() {
    // all error codes are prototype properties
    expect(Kapow.prototype).to.contain.all.keys(errorCodes);
    // all prototype properties are functions
    expect(_.functions(Kapow.prototype)).to.deep.equal(Object.keys(Kapow.prototype));
  });
});


test/unit/lib/wrap.js
const expect = require('chai').expect;

const wrap = require('../../../lib/wrap');
const RequestError = require('../../../lib/request_error');

describe('wrap', function() {
  it('should default to a RequestError', function() {
    expect(wrap()).to.be.instanceof(RequestError);
  });

  it('should set error to the passed-in error argument', function() {
    expect(wrap(new TypeError())).to.be.instanceof(TypeError);
  });

  it('should default to httpStatus: 400', function() {
    expect(wrap().httpStatus).to.equal(400);
  });

  it('should set httpStatus to the passed-in argument', function() {
    var code = 500;
    expect(wrap(null, code, null).httpStatus).to.equal(code);
  });

  it('should default to message: "Client Error"', function() {
    expect(wrap().message).to.equal('Client Error');
  });

  it('should set message to the passed-in argument', function() {
    var message = 'My Message';
    expect(wrap(null, null, message).message).to.equal(message);
  });

  it('should default to title: "Bad Request"', function() {
    expect(wrap().title).to.equal('Bad Request');
  });
});


lib/error_codes.js
module.exports = {
  // RFC 7231 - http://tools.ietf.org/html/rfc7231#page-47
  '400': 'Bad Request',
  '401': 'Unauthorized',
  '402': 'Payment Required',
  '403': 'Forbidden',
  '404': 'Not Found',
  '405': 'Method Not Allowed',
  '406': 'Not Acceptable',
  '407': 'Proxy Authentication Required',
  '408': 'Request Timeout',
  '409': 'Conflict',
  '410': 'Gone',
  '411': 'Length Required',
  '412': 'Precondition Failed',
  '413': 'Request Entity Too Large',
  '414': 'Request-URI Too Long',
  '415': 'Unsupported Media Type',
  '416': 'Requested Range Not Satisfiable',
  '417': 'Expectation Failed',
  '426': 'Upgrade Required',

  // RFC 2324 :-P - http://tools.ietf.org/html/rfc2324
  '418': 'I\'m A Teapot',

  // RFC 4918 - https://tools.ietf.org/html/rfc4918#page-78
  '422': 'Unprocessable Entity',
  '423': 'Locked',
  '424': 'Failed Dependency',

  // RFC 6585 - http://tools.ietf.org/html/rfc6585
  '428': 'Precondition Required',
  '429': 'Too Many Requests',
  '431': 'Header Fields Too Large',

  // Internet-Draft -
  // http://tools.ietf.org/html/draft-tbray-http-legally-restricted-status-05
  '451': 'Unavailable For Legal Reasons',

  // RFC 7231 - http://tools.ietf.org/html/rfc7231#page-47
  '500': 'Internal Server Error',
  '501': 'Not Implemented',
  '502': 'Bad Gateway',
  '503': 'Service Unavailable',
  '504': 'Gateway Timeout',
  '505': 'HTTP Version Not Supported',

  // RFC 2295 - https://tools.ietf.org/html/rfc2295#page-25
  '506': 'Variant Also Negotiates',

  // RFC 4918 - https://tools.ietf.org/html/rfc4918#page-78
  '507': 'Insufficient Storage',

  // RFC 5842 - https://tools.ietf.org/html/rfc5842#page-31
  '508': 'Loop Detected',

  // RFC 2774 - https://tools.ietf.org/html/rfc2774#page-11
  '510': 'Not Extended',

  // RFC 6585 - http://tools.ietf.org/html/rfc6585
  '511': 'Network Authentication Required'
};
